### ISSUE TITLE
Fix a race found in unit tests

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#15767]
This is probably introduced after the RPC tcp stats but it is not clear
how it is related yet

```
=== RUN   TestAPIListObjectPartsHandlerPreSign
WARNING: DATA RACE
Read at 0x00c0364bf4d0 by goroutine 1044:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.18.6/x64/src/runtime/panic.go:436 +0x32
  github.com/minio/minio/cmd.newlockRESTClient.func1()
      /home/runner/work/minio/minio/cmd/lock-rest-client.go:163 +0x104
  github.com/minio/minio/internal/rest.(*Client).MarkOffline.func1()
      /home/runner/work/minio/minio/internal/rest/client.go:359 +0xb5

Previous write at 0x00c0364bf4d0 by goroutine 3077:
  github.com/minio/minio/internal/rest.setupReqStatsUpdate.func1()
      /home/runner/work/minio/minio/internal/rest/rpc-stats.go:61 +0x4d
  net.(*sysDialer).dialSingle()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/dial.go:573 +0x173
  net.(*sysDialer).dialSerial()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/dial.go:551 +0x471
  net.(*Dialer).DialContext()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/dial.go:428 +0xabb
  github.com/minio/minio/internal/http.NewInternodeDialContext.func1()
      /home/runner/work/minio/minio/internal/http/dial_linux.go:68 +0xe5
  github.com/minio/minio/internal/http.DialContextWithDNSCache.func1()
      /home/runner/work/minio/minio/internal/http/dial_dnscache.go:63 +0x343
  net/http.(*Transport).dial()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1169 +0x269
  net/http.(*Transport).dialConn()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1607 +0xd57
  net/http.(*Transport).dialConnFor()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1449 +0x13a
  net/http.(*Transport).queueForDial.func1()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1418 +0x47

Goroutine 1044 (running) created at:
  github.com/minio/minio/internal/rest.(*Client).MarkOffline()
      /home/runner/work/minio/minio/internal/rest/client.go:354 +0x364
  github.com/minio/minio/internal/rest.(*Client).Call()
      /home/runner/work/minio/minio/internal/rest/client.go:252 +0x5a4
  github.com/minio/minio/cmd.(*lockRESTClient).callWithContext()
      /home/runner/work/minio/minio/cmd/lock-rest-client.go:65 +0x134
  github.com/minio/minio/cmd.(*lockRESTClient).restCall()
      /home/runner/work/minio/minio/cmd/lock-rest-client.go:97 +0x21e
  github.com/minio/minio/cmd.(*lockRESTClient).RLock()
      /home/runner/work/minio/minio/cmd/lock-rest-client.go:111 +0x19b
  github.com/minio/minio/cmd.TestLockRESTlient()
      /home/runner/work/minio/minio/cmd/lock-rest-client_test.go:40 +0x15a
  testing.tRunner()
      /opt/hostedtoolcache/go/1.18.6/x64/src/testing/testing.go:1439 +0x213
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.18.6/x64/src/testing/testing.go:1486 +0x47

Goroutine 3077 (running) created at:
  net/http.(*Transport).queueForDial()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1418 +0x172
  net/http.(*Transport).getConn()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:1372 +0x893
  net/http.(*Transport).roundTrip()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/transport.go:581 +0xcc4
  net/http.(*Transport).RoundTrip()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/roundtrip.go:17 +0x36
  net/http.send()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/client.go:252 +0x93e
  net/http.(*Client).send()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/client.go:176 +0x157
  net/http.(*Client).do()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/client.go:725 +0x1064
  net/http.(*Client).Do()
      /opt/hostedtoolcache/go/1.18.6/x64/src/net/http/client.go:593 +0x4f1
  github.com/minio/minio/internal/rest.(*Client).Call()
      /home/runner/work/minio/minio/internal/rest/client.go:246 +0x4c8
  github.com/minio/minio/cmd.newlockRESTClient.func1()
      /home/runner/work/minio/minio/cmd/lock-rest-client.go:163 +0x104
  github.com/minio/minio/internal/rest.(*Client).MarkOffline.func1()
      /home/runner/work/minio/minio/internal/rest/client.go:359 +0xb5
==================
    testing.go:1312: race detected during execution of test
--- FAIL: TestAPIListObjectPartsHandlerPreSign (91.89s)

```

**Comment by harshavardhana** [comment]
```
        var dialStart, dialEnd time.Time

        trace := &httptrace.ClientTrace{
                ConnectStart: func(network, addr string) {
                        dialStart = time.Now()
                },
                ConnectDone: func(network, addr string, err error) {
                        if err == nil {
                                dialEnd = time.Now()
                        }
                },
        }

        return req.WithContext(httptrace.WithClientTrace(req.Context(), trace)), func() {
                if !dialStart.IsZero() {
                        if dialEnd.IsZero() {
                                atomic.AddUint64(&globalStats.tcpDialErrs, 1)
                        } else {
                                atomic.AddUint64(&globalStats.tcpDialCount, 1)
                                atomic.AddUint64(&globalStats.tcpDialTotalDur, uint64(dialEnd.Sub(dialStart)))
                        }
                }
        }
}
```

This kind of code is racy `dialStart` is parallelly read @vadmeste

**Comment by harshavardhana** [comment]
use this https://github.com/uber-go/atomic/blob/v1.10.0/time.go#L39

**Comment by vadmeste** [comment]
My suspicion is that is a false race detection. My theory is that Golang net/http library creates a goroutine which will update dialStart and dialEnd. In parallel, we also have our internal rest lock/storage/peer calls which can be called in a goroutine and will read dialStart and dialEnd to update our statistics. However, I don't think golang net/http will mess this up and crash.

I will use atomic anyway to update

**Comment by klauspost** [comment]
> My suspicion is that is a false race detection. 

False positives are not possible. False negatives are.

**Comment by klauspost** [comment]
I don't see the connection to the code above, though.

**Comment by klauspost** [comment]
@vadmeste I think the request is being re-used.

Storing it atomically is not going to fix the race. Well, it will look good, but the stats will be wrong.

**Comment by vadmeste** [comment]
@klauspost I found a better race report in the PR description here, https://github.com/minio/minio/pull/15770

MinIO code clearly races with the internal golang code

**Comment by klauspost** [comment]
@vadmeste I think the root cause is that the `*http.Request` is re-used, so when you are recording the stats another one can update the values.

**Comment by klauspost** [comment]
(or it tries to dial multiple times, something like that)

**Comment by harshavardhana** [comment]
fixed via #15770
